# Project 3: Skiing vs Snowboarding
## Problem Statement
Mad River Glen is a ski resort in Vermont that has stood their ground on keeping snowboarders off the mountain. Year after year they are pressured by snowboarders to allow them on the mountain. Many proclaim, "there is no difference between snowboarding and skiing!" They have asked me, the data scientist on the crew, to prove that there is a difference between skiing and snowboarding using Reddit and natural language processing. 
## Scraping

In [1]:
# Imports
import numpy as np
import pandas as pd
import praw
import time
import requests
import datetime as dt

In [2]:
# Creating a reddit instance
# reddit = praw.Reddit(
#     client_id = 'Ca5paMO6FqJDUw',
#     client_secret = 'JFv1OgKfet5Ud8ePxq0nRz4fnfg',
#     username = 'lelandroberts97',
#     password = 'Yankees01!',
#     user_agent = 'lelandroberts'
# )

In [3]:
# Creating our subreddit objects
# r_ski = reddit.subreddit('skiing')
# r_snowboard = reddit.subreddit('snowboarding')

In [4]:
# Creating lists for the attributes we care about: title, selftext, id, created_utc, author
# ski_titles, ski_selftexts, ski_ids, ski_createds, ski_authors = [], [], [], [], []
# board_titles, board_selftexts, board_ids, board_createds, board_authors = [], [], [], [], []

### PRAW: r/skiing

In [5]:
# Populating our sets of documents
# for submission in r_ski.hot(limit=1000):
#     time.sleep(0.5)
#     if submission.is_self == True: # Getting posts with text only
#         ski_ids.append(submission.id)
#         ski_titles.append(submission.title)
#         ski_selftexts.append(submission.selftext)
#         ski_createds.append(submission.created)
#         ski_authors.append(submission.author)
#     if len(ski_ids) >= 500:
#         break

In [6]:
# Creating a data frame with the ski documents   
# ski = pd.DataFrame({
#     'id': ski_ids,
#     'title': ski_titles,
#     'text': ski_selftexts,
#     'created': ski_createds,
#     'author': ski_authors,
# })

In [7]:
# Adding a column for class
# ski['class'] = 1

In [8]:
# ski.head() # Taking a look at the data frame

In [9]:
# ski.shape # Checking the size to make sure we have enough

### PRAW: r/snowboarding

In [10]:
# for submission in r_snowboard.hot(limit=1000):
#     time.sleep(0.5)
#     if submission.is_self == True: # Getting posts with text only
#         board_ids.append(submission.id)
#         board_titles.append(submission.title)
#         board_selftexts.append(submission.selftext)
#         board_createds.append(submission.created)
#         board_authors.append(submission.author)
#     if len(board_ids) >= 500:
#         break

In [11]:
# Creating a data frame with the snowboard documents    
# board = pd.DataFrame({
#     'id': board_ids,
#     'title': board_titles,
#     'text': board_selftexts,
#     'created': board_createds,
#     'author': board_authors,
# })

In [12]:
# Adding a column for class
# board['class'] = 0

In [13]:
# board.head() # Taking a look at the data frame

In [14]:
# board.shape # Checking the size to make sure we have enough

## Export

In [15]:
# Saving data frames to csv files
# ski.to_csv('../Data/ski.csv', index=False)
# board.to_csv('../Data/board.csv', index=False)

## Additional Scraping with Requests
I will use the requests library to see if I can get additional data. Also, as the week progresses, I will be running similar code as above for the 50 most frequent posts and add them to my dataframe.

In [89]:
# Reading in my original data frame
# skiing_0 = pd.read_csv('../Data/ski.csv')
# boarding = pd.read_csv('../Data/board.csv')

In [47]:
# Defining urls and user agent
# skiing_base_url = 'https://www.reddit.com/r/skiing.json'
# boarding_base_url = 'https://www.reddit.com/r/snowboarding.json'

# headers={'User-agent':'lrob5178'}

### Requests: r/skiing

In [51]:
# Using the requests library to scrape reddit for documents in r/skiing
# skiing_post=[]
# after=None
# for i in range(50):
#     if after==None:
#         skiing_params={}
#     else: skiing_params={'after':after}
#     skiing_res=requests.get(skiing_base_url,headers=headers,params=skiing_params)
#     if skiing_res.status_code==200:
#         skiing_json=skiing_res.json()
#         no_blanks=[c['data'] for c in skiing_json['data']['children'] if len(c['data']['selftext'])>10]
#         skiing_post.extend(no_blanks)
#         after=skiing_json['data']['after']
#     else:
#         print(skiing_res.status_code)
#         break
#     time.sleep(1)
#     print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [113]:
# Creating a data frame with only the features we are interested in: id, title, author, selftext, and created_utc
# skiing = pd.DataFrame(skiing_post).loc[:, ['id', 'title', 'author', 'selftext', 'created_utc']]

In [114]:
# skiing['class'] = 1 # Creating a column for the class (our predictor variable)

In [116]:
# skiing.shape # Checking to see how many documents we got

(206, 6)

In [6]:
# Renaming columns to match original data frame
# col_names = {
#     'selftext': 'text',
#     'created_utc': 'created'
# }

In [ ]:
# skiing = skiing.rename(columns=col_names)

In [118]:
# Adding the data scaped using the requests library to the original data frame obtained by using PRAW
# skiing_2 = pd.concat([skiing_0, skiing])

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


In [119]:
# skiing_2.shape # Checking the size

(426, 6)

In [120]:
# skiing_2.drop_duplicates(subset=['text', 'title', 'author']).shape # Checking the size without duplicates

(238, 6)

#### Note:
I did not obtain many additional unique documents from using the requests library. I originally had 220 documents, and, after adding the new documents and dropping the duplicates, I had 238 documents. This makes me think there is a limit to how far you can go back. Instead of just using r/skiing and r/snowboarding, I will try is scraping r/ski and r/snowboard as well, as these are very similar subreddits. For our purposes, there shouldn't be an issue combining them.

### Requests: r/ski

In [86]:
# Using the requests library to scrape reddit for documents in r/ski
# ski_base_url = 'https://www.reddit.com/r/ski.json'
# ski_post=[]
# after=None
# for i in range(50):
#     if after==None:
#         ski_params={}
#     else: ski_params={'after':after}
#     ski_res=requests.get(ski_base_url,headers=headers,params=ski_params)
#     if ski_res.status_code==200:
#         ski_json=ski_res.json()
#         no_blanks=[c['data'] for c in ski_json['data']['children'] if len(c['data']['selftext'])>10]
#         ski_post.extend(no_blanks)
#         after=ski_json['data']['after']
#     else:
#         print(ski_res.status_code)
#         break
#     time.sleep(1)
#     print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [122]:
# Creating a data frame with only the features we are interested in: id, title, author, selftext, and created_utc
# ski = pd.DataFrame(ski_post).loc[:, ['id', 'title', 'author', 'selftext', 'created_utc']]
# board['class'] = 1

In [123]:
# ski.shape # Checking the shape

(590, 5)

In [124]:
# ski = ski.rename(columns=col_names) # Ranaming columns to match original data frame

In [126]:
# ski = ski.drop_duplicates(subset=['text', 'title', 'author']) # Dropping duplicates

In [127]:
# ski_2 = pd.concat([skiing_0, ski]) # Combining r/ski and r/skiing data frames

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [128]:
# ski_2.shape # Checking how many documents we have

(675, 6)

675 is a much better number of documents to work with!

### Requests Library: r/snowboard

In [62]:
# Using the requests library to scrape reddit for documents in r/snowboard
# board_base_url = 'https://www.reddit.com/r/snowboard.json'
# board_post=[]
# after=None
# for i in range(50):
#     if after==None:
#         board_params={}
#     else: board_params={'after':after}
#     board_res=requests.get(board_base_url,headers=headers,params=board_params)
#     if board_res.status_code==200:
#         board_json=board_res.json()
#         no_blanks=[c['data'] for c in board_json['data']['children'] if len(c['data']['selftext'])>10]
#         board_post.extend(no_blanks)
#         after=board_json['data']['after']
#     else:
#         print(board_res.status_code)
#         break
#     time.sleep(1)
#     print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [131]:
# Creating a data frame with only the features we are interested in: id, title, author, selftext, and created_utc
# board = pd.DataFrame(board_post).loc[:, ['id', 'title', 'author', 'selftext', 'created_utc']]
# board['class'] = 0

In [132]:
# board.shape # Checking to see how many documents we got

(495, 6)

In [133]:
# board = board.rename(columns=col_names) # Renaming the columns to match the original data frame

In [136]:
# board = board.drop_duplicates(subset=['text', 'title', 'author']) # Dropping the duplicates

In [137]:
# board_2 = pd.concat([boarding, board]) # Combining the r/snowboard and r/snowboarding data frames

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [138]:
# board_2.shape # Checking how many documents we have

(221, 6)

221 documents is still not enough to work with, especially considering how many we have for the skiing class. I will have to try something else to get more data for the snowboarding class.

## Export

In [140]:
# Saving data frames to csv files
# ski_2.to_csv('../Data/ski_2.csv', index=False)
# board_2.to_csv('../Data/board_2.csv', index=False)

## Additional Scraping with Pushshift.io

In [3]:
# Function from Raffy
def get_date(created):
    # get the date of post
    return dt.date.fromtimestamp(created)

def query_pushshift(subreddit, kind='submission', skip=5, times=50, 
                    subfield = ['title', 'selftext', 'subreddit', 'created_utc', 'author', 'is_self']):
    '''
    subreddit: The subreddit you want to scrape (required)
    kind = 'submission': Scrape the specified post kind
    skip = 5: For each iteration of `times`, skip `skip` number of days
    times = 50: The number of iterations to make
    subfield = long list to type: List of information that is needed
    '''
    # get the base url that contains information I want to scrape where 'kind' are all submitted posts
    # and 'subreddit' is the specified subreddit. Get 500 posts.
    stem = f"https://api.pushshift.io/reddit/search/{kind}/?subreddit={subreddit}&size=500"
    # instantiate list to contain 
    mylist = []
    # scrape posts from the subreddit 'times' times
    for x in range(1, times + 1):
        # Get posts 'skip' * 'x' days ago
        URL = f"{stem}&after={skip * x}d"
        print(URL)
        # Scrape URL
        response = requests.get(URL)
        # Give me an AssertionError if status code not 200
        assert response.status_code == 200
        # Of the HTML scraped, take the values of 'data'
        the_json=response.json()
        no_blanks=[c for c in the_json['data'] if ('selftext' in c.keys()) and len(c['selftext'])>10]
        # turn the data into a dataframe
        df = pd.DataFrame.from_dict(no_blanks)
        # append the dataframe to mylist
        mylist.append(df)
        # wait to not overrun Reddit's resources
        time.sleep(0.5)
    # concatenate the dataframes together as one large dataframe, full
    full = pd.concat(mylist, sort=False)
    if kind == "submission":
        # take all speficied data
        full = full[subfield]
        # drop duplicate rows
        full = full.drop_duplicates()
        full = full.loc[full['is_self'] == True]
    # date the the post was... posted
    _timestamp = full["created_utc"].apply(get_date)
    full['timestamp'] = _timestamp
    print(full.shape)
    return full

### Pushshift.io: r/skiing

In [16]:
# Getting data from r/skiing
ski_push = query_pushshift('skiing')

https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=5d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=10d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=15d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=20d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=25d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=30d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=35d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=40d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=45d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=50d
https://api.pushshift.io/reddit/search/submission/?subreddit=skiing&size=500&after=55d
https://api.pushshift.io/reddit/search/submi

In [20]:
ski_push['class'] = 1 # Creating a column for our predictor variable

In [17]:
ski_push = ski_push.rename(columns=col_names) # Renaming the columns to match the original data frame

In [22]:
ski_push.shape[0] - ski_push.drop_duplicates(subset=['title', 'text', 'author']).shape[0] # Checking for duplicates

5

In [24]:
ski_push = ski_push.drop_duplicates(subset=['text', 'title', 'author']) # Dropping the duplicates

In [26]:
ski_push.head() # Viewing the data frame

title  \
0                                 First set of skis.   
1                        Eastern Pennsylvania Skiing   
2  OK Marksman, Bentchetler 100 for a solid daily...   
3  Hank Bilous Massive Transfer &amp; Full Run | ...   
4                                      Niseko Danger   

                                                text subreddit     created  \
0  So I am thinking about getting my first skis (...    skiing  1579818807   
1  Looking for any advice at all. Or buddies! Goi...    skiing  1579822807   
2                         Thanks for the help y’all!    skiing  1579823764   
3  [https://www.newschoolers.com/videos/watch/951...    skiing  1579824890   
4  Just trying to get a general feel for how dang...    skiing  1579827729   

              author  is_self   timestamp  class  
0              D3470     True  2020-01-23      1  
1          potatoes6     True  2020-01-23      1  
2     mrrichmahogany     True  2020-01-23      1  
3      jurassisaurus     True  2020-01-23      1  
4  IndoorSurvivalist     True  2020-01-23      1

### Pushshift.io: r/snowboarding

In [15]:
# Getting data from r/skiing
board_push = query_pushshift('snowboarding')

https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=5d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=10d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=15d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=20d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=25d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=30d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=35d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=40d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=45d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboarding&size=500&after=50d
https://api.pushshift.io/reddit/search/submission/?subreddit=snowboardi

In [21]:
board_push['class'] = 0 # Creating a column for our predictor variable

In [19]:
board_push = board_push.rename(columns=col_names) # Renaming the columns to match the original data frame

In [23]:
board_push.shape[0] - board_push.drop_duplicates(subset=['title', 'text', 'author']).shape[0] # Checking for duplicates

5

In [25]:
board_push = board_push.drop_duplicates(subset=['text', 'title', 'author']) # Dropping the duplicates

In [27]:
board_push.head() # Viewing the data frame

title  \
0             Buying a board, need opinions on size.   
1                                   Beginner advice!   
2  Burton snowboarding jacket sizing for a small guy   
3  New Snowboarder here, what’s the deal with the...   
4                  Do FORUM Boards have any value???   

                                                text     subreddit  \
0  I want to get back into snowboarding, havent d...  snowboarding   
1  So a little context, I used to ski, went ridin...  snowboarding   
2  I am no big guy by American standards: 5'6", a...  snowboarding   
3  So my wife and I are just getting into snowboa...  snowboarding   
4                      like is it worth keeping one?  snowboarding   

      created          author  is_self   timestamp  class  
0  1579830441          tfr737     True  2020-01-23      0  
1  1579830507         Mcapp-1     True  2020-01-23      0  
2  1579835749      bmw-fanboy     True  2020-01-23      0  
3  1579836431     GoingDigitl     True  2020-01-23      0  
4  1579849635  Contentclicker     True  2020-01-24      0

#### Note:
I was able to get nearly 4000 original documents from both r/skiing and r/snowboarding using pushshift.io. I will use this data going forward.

## Export

In [30]:
# Saving data frames to csv files
ski_push.to_csv('../Data/skiing.csv', index=False)
board_push.to_csv('../Data/snowboarding.csv', index=False)